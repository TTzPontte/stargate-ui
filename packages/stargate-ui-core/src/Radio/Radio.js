import React, { useRef, useState } from 'react';
import PropTypes from "prop-types";

import { useStyles } from '@pontte/stargate-ui-styles';
import { InputLabel } from '@pontte/stargate-ui-core';
import Typography from '../Typography';
import Factory from '../Factory';

const styles = (theme) => {
  const {
    palette,
    radius,
    mode,
  } = theme;

  const { setLightness } = palette;
  const getColor = (color) => (
    palette?.[color][mode].color
  );

  const radio = {
    position: 'absolute',
    width: 0,
    height: 0,
    userSelect: 'none',
    opacity: 0,
  };

  const radioChecked = {
    cursor: 'pointer',
    '&:before': {
      content: '""',
      display: 'inline-flex',
      width: 20,
      height: 20,
      borderRadius: radius(3),
      backgroundColor: palette.lighter,
      margin: 0,
      border: ({color}) => (
        [[1, 'solid', getColor(color)]]
      ),
    },
    '&:after': {
      content: '""',
      display: 'inline-flex',
      '$radio:checked ~ &': {
        bottom: 6,
        right: 16,
        width: 12,
        height: 12,
        position: 'relative',
        borderRadius: radius(2),
        border: (props) => {
          const {
            disabled,
            readonly,
            color,
          } = props;

          if (disabled || readonly) {
            return [['solid', setLightness(.95, getColor(color))]];
          }

          return [['solid', (color !== 'default' ? getColor(color) : getColor('success'))]];
        },
        backgroundColor: (props) => {
          const {
            disabled,
            readonly,
            color,
          } = props;

          if (disabled || readonly) {
            return setLightness(.95, getColor(color));
          }

          return color !== 'default' ? getColor(color) : getColor('success');
        },
      },
    },
  };

  const radioWrapper = {
    width: 'fit-content',
    display: 'flex',
  }

  return {
    radio,
    radioChecked,
    radioWrapper,
  };
};

const Radio = (props) => {
  const {
    disabled,
    readonly,
    label,
    color = 'default',
    checked: defaultValue = false,
    onChange = () => {},
    ...factoryProps
  } = props;

  const [
    {
      radio: classRadio,
      radioChecked: classRadioChecked,
      radioWrapper: classRadioWrapper,
    }
  ] = useStyles(styles, {
    disabled,
    readonly,
    color,
  });

  const [checked, setChecked] = useState(defaultValue);
  const inputRef = useRef();

  const handleClick = () => {
    if (disabled) {
      return;
    }

    setChecked(!checked);
    onChange({ checked });
  }

  return (
    <Factory className={classRadioWrapper}>
      <InputLabel>
        <Factory
          ref={inputRef}
          element="input"
          type="radio"
          className={classRadio}
          checked={checked}
          onChange={handleClick}
          aria-label="Radio Button"
          />

        <Factory
          element="span"
          className={classRadioChecked}
          marginX={1}
          onClick={handleClick}
          {...factoryProps}
        />

        <Typography
          variant="body"
          element="span"
          children={label}
        />
      </InputLabel>
    </Factory>
  );
};

Radio.displayName = 'Radio';

Radio.propTypes = {
  /**
   * Add disabled style.
   * @default undefined
   */
  disabled: PropTypes.bool,
  /**
   * Add checked style.
   * @default false
   */
  checked: PropTypes.bool.isRequired,
  /**
   * Add readonly style.
   * @default undefined
   */
  readonly: PropTypes.bool,
  /**
   * Add label style.
   * @default undefined
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  /**
   * Add type style.
   * @default undefined
   */
  type: PropTypes.string,
  /**
   * Add color style.
   * @default default
   */
  color: PropTypes.oneOf([
    'default',
    'success',
    'warning',
    'info',
    'error'
  ]),
  /**
   * Trigger when element is clicked.
   * @default Function
   */
  onChange: PropTypes.func,
};

/**
 * Add @property {object} factoryProps made available properties information
 * for Props in the Storybook but do not use as major define for default properties.
 */
Radio.defaultProps = {
  color: 'default',
  checked: false,
  onChange: () => {},
};

export default Radio;
