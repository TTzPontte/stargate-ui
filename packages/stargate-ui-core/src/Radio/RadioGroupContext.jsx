import React from 'react';

const RadioGroupContext = React.createContext({});

RadioGroupContext.displayName = 'RadioGroupContext';

export default RadioGroupContext;
