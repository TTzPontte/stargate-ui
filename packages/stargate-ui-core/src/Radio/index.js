export { default } from './RadioGroup';
export { default } from './Radio';
