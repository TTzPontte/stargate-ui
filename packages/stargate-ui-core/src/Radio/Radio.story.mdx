import {
  Meta,
  Canvas,
  Status,
  Story,
  withAction,
  Props,
} from '@pontte/stargate-ui-playground';

import Radio from './Radio';

<Meta title="Core/Form Elements/Radio" component={Radio} />

# Radio Button

<Status type="rc" />

## Basic Usage

<Canvas>
  <Story name="Basic Usage" parameters={{
    jest: ['Radio.spec.jsx'],
  }}>
    {
      () => (
        <Radio
          label="Radio Button"
          checked={false}
        />
      )
    }
  </Story>
</Canvas>

<Props component={Radio} />

### Default Radio Button

<Canvas>
  <Radio
    label="Default Checked"
    checked
  />
  <Radio
    label="Default Checked Disabled"
    checked
    disabled
  />
</Canvas>

### Colors State

<Canvas>
  <Radio
    label
    color="success"
    checked={false}
  />
  <Radio
    color="warning"
    checked
  />
  <Radio
    color="error"
    checked={false}
  />
  <Radio
    color="info"
    checked
  />
</Canvas>
