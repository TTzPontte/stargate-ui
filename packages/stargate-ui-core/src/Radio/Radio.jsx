import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import { useStyles } from '@pontte/stargate-ui-styles';
import Label from '../Label';
import Factory from '../Factory';
import Typography from '../Typography';
import RadioGroupContext from './RadioGroupContext';

const styles = (theme) => {
  const {
    palette,
    radius,
    mode,
  } = theme;

  const radio = {
    position: 'absolute',
    width: 0,
    height: 0,
    userSelect: 'none',
    opacity: 0,
  };

  const radioDecoration = {
    cursor: 'pointer',
    width: 20,
    height: 20,
    display: 'inline-block',
    position: 'relative',
    borderRadius: radius(3),
    background: palette.lighter,
    margin: 0,
    overflow: 'hidden',
    border: [
      1,
      'solid',
      palette.default[mode].color,
    ],
    '$radio:checked ~ &:after': {
      content: '""',
      display: 'inline-flex',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
      width: '60%',
      height: '60%',
      position: 'absolute',
      borderRadius: radius(2),
    },
    '$radio:checked:disabled ~ &:after': {
      background: palette.setLightness(.90, palette.default[mode].color),
    },
  };

  const radioDecorationSuccess = {
    '$radio:checked ~ &:after': {
      background: palette.success[mode].color,
    },
  };

  const radioDecorationWarning = {
    '$radio:checked ~ &:after': {
      background: palette.warning[mode].color,
    },
  };

  const radioDecorationError = {
    '$radio:checked ~ &:after': {
      background: palette.error[mode].color,
    },
  };

  return {
    radio,
    radioDecoration,
    radioDecorationSuccess,
    radioDecorationWarning,
    radioDecorationError,
  };
};

const Radio = React.forwardRef((props, ref) => {
  const {
    label,
    value,
    color: defaultColor,
    checked: defaultChecked,
    disabled: defaultDisabled,
    readOnly: defaultReadOnly,
    ...factoryProps
  } = props;

  const innerRef = React.useRef(ref);

  const {
    name,
    onChange,
    color: radioGroupColor,
    value: radioGroupValue,
    disabled: radioGroupDisabled,
    readOnly: radioGroupReadonly,
  } = React.useContext(RadioGroupContext);

  const checked = value && radioGroupValue ? value === radioGroupValue : defaultChecked;
  /**
   * Local properties has priority over @func RadioGroupContext.
   */
  const color = defaultColor || radioGroupColor || 'success';
  const disabled = defaultDisabled ?? radioGroupDisabled;
  const readOnly = defaultReadOnly ?? radioGroupReadonly;

  const handleChange = (event) => {
    if (disabled || readOnly) {
      event.preventDefault();
      return;
    }
    onChange(event, [value]);
  };

  const [{ radio: classRadio, ...classes }] = useStyles(styles);
  const classRadioDecoration = clsx(
    classes.radioDecoration,
    {
      [classes.radioDecorationSuccess]: color === 'success',
      [classes.radioDecorationError]: color === 'error',
      [classes.radioDecorationWarning]: color === 'warning',
    },
  );

  return (
    <Factory display="flex">
      <Label>
        <Factory
          {...factoryProps}
          disabled={disabled}
          readOnly={readOnly}
          checked={checked}
          ref={innerRef}
          element="input"
          type="radio"
          className={classRadio}
          onChange={handleChange}
          value={value}
          disabled={disabled}
          name={name}
        />

        <Factory
          element="span"
          className={classRadioDecoration}
          marginX={1}
        />

        <Typography
          element="span"
          variant="body"
          children={label}
          gutter={0}
        />
      </Label>
    </Factory>
  );
});

Radio.displayName = 'Radio';

Radio.propTypes = {
  /**
   * Disables button and add disabled CSS style.
   *
   * **@default** `undefined`
   */
  disabled: PropTypes.bool,
  /**
   * Add readonly CSS style.
   *
   * **@default** `undefined`
   */
  readonly: PropTypes.bool,
  /**
   * Add a label description for each radio.
   *
   * **@default** `undefined`
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]).isRequired,
  /**
   * Add color style.
   *
   * **@default** `success`
   */
  color: PropTypes.oneOf([
    'success',
    'warning',
    'error'
  ]),
};

export default Radio;
