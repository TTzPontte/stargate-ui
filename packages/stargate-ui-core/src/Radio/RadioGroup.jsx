import React from 'react';
import PropTypes from 'prop-types';

import RadioGroupContext from './RadioGroupContext';

const RadioGroup = (props) => {
  const {
    children,
    name,
    value,
    disabled,
    readOnly,
    color,
    onChange = () => {},
  } = props;

  const handleChange = React.useCallback(onChange, [onChange]);
  const context = React.useMemo(() => ({
    name,
    value,
    disabled,
    readOnly,
    color,
    onChange: handleChange,
  }), [
    name,
    value,
    disabled,
    readOnly,
    color,
    handleChange,
  ]);

  return (
    <RadioGroupContext.Provider
      value={context}
      children={children}
    />
  );
};

RadioGroup.displayName = 'RadioGroup';

RadioGroup.propTypes = {};

export default RadioGroup;
