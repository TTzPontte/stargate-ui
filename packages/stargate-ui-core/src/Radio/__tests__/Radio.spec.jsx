import React from 'react';

import Radio from '../Radio';

describe('Radio', () => {
  it('should render properly', () => {
    const { container } = render(
      <Radio label="Radio Button"/>
    );

    expect(container).toMatchSnapshot();
  });
});
