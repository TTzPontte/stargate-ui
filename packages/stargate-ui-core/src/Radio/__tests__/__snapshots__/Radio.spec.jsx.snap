// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should render properly 1`] = `
<div>
  <input
    class="stargate-ui-factory-0-2-4 stargate-ui-factory-d0-0-2-6 stargate-ui-radio-0-2-1"
    type="radio"
  />
  <span
    class="stargate-ui-factory-0-2-8 stargate-ui-factory-d0-0-2-10 stargate-ui-radioChecked-0-2-2 stargate-ui-radioChecked-d0-0-2-3"
  />
  <span
    class="stargate-ui-factory-0-2-19 stargate-ui-factory-d0-0-2-21 stargate-ui-typography-0-2-12 stargate-ui-typography-d0-0-2-16 stargate-ui-typographyFluid-0-2-15 stargate-ui-typographyFluid-d2-0-2-18"
  >
    Radio Button
  </span>
</div>
`;
